File: backend/app/schemas/rain_cell.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.common import LatLng

TrackState = Literal["NEW", "TRACKING", "LOST", "EXPIRED"]
RainCellTrackStatus = Literal["ok", "partial", "unavailable"]


class RainCellTrackRequest(BaseModel):
    geometry: list[LatLng] = Field(..., min_length=2)
    buffer_km: float | None = Field(default=None, ge=1, le=300)


class CellIntensityOut(BaseModel):
    min: float | None = None
    max: float | None = None
    mean: float | None = None


class CellBoundsOut(BaseModel):
    north: float
    south: float
    east: float
    west: float


class RainCellOut(BaseModel):
    id: str
    timestamp: str
    centroid: LatLng
    area_km2: float | None = None
    intensity: CellIntensityOut | None = None
    bounds: CellBoundsOut | None = None


class CellMotionOut(BaseModel):
    speed_kmh: float | None = None
    bearing_degrees: float | None = None
    from_point: LatLng | None = None
    to_point: LatLng | None = None
    confidence: float | None = None


class TrackedRainCellOut(BaseModel):
    id: str
    state: TrackState
    current: RainCellOut
    history: list[RainCellOut]
    motion: CellMotionOut | None = None
    distance_to_route_km: float | None = None
    missed_frames: int = 0


class RainCellTrackResponse(BaseModel):
    status: RainCellTrackStatus
    frames_used: int
    cells: list[TrackedRainCellOut]
    message: str | None = None

File: frontend/app/types/rainCell.ts
import type { LatLng } from "~/types/routeWeather"

export type TrackState = "NEW" | "TRACKING" | "LOST" | "EXPIRED"
export type RainCellTrackStatus = "ok" | "partial" | "unavailable"

export type CellIntensity = {
  min?: number | null
  max?: number | null
  mean?: number | null
}

export type CellBounds = {
  north: number
  south: number
  east: number
  west: number
}

export type RainCell = {
  id: string
  timestamp: string
  centroid: LatLng
  area_km2?: number | null
  intensity?: CellIntensity | null
  bounds?: CellBounds | null
}

export type CellMotion = {
  speed_kmh?: number | null
  bearing_degrees?: number | null
  from_point?: LatLng | null
  to_point?: LatLng | null
  confidence?: number | null
}

export type TrackedRainCell = {
  id: string
  state: TrackState
  current: RainCell
  history: RainCell[]
  motion?: CellMotion | null
  distance_to_route_km?: number | null
  missed_frames: number
}

export type RainCellTrackResponse = {
  status: RainCellTrackStatus
  frames_used: number
  cells: TrackedRainCell[]
  message?: string | null
}

export type RainCellTrackRequest = {
  geometry: LatLng[]
  buffer_km?: number | null
}
